patches/minecraft/net/minecraft/advancements/Advancement.java.patch

```diff
@@ -1,5 +1,36 @@
 --- a/net/minecraft/advancements/Advancement.java
 +++ b/net/minecraft/advancements/Advancement.java
+@@ -230,12 +_,30 @@
+          }
+       }
+ 
++      @Deprecated //Forge: Use Advancement#save(Consumer<Advancement>, ResourceLocation, ExistingFileHelper)
+       public Advancement func_203904_a(Consumer<Advancement> p_203904_1_, String p_203904_2_) {
+          Advancement advancement = this.func_192056_a(new ResourceLocation(p_203904_2_));
+          p_203904_1_.accept(advancement);
+          return advancement;
+       }
+ 
++      public Advancement save(Consumer<Advancement> consumer, ResourceLocation id, net.minecraftforge.common.data.ExistingFileHelper fileHelper) {
++         Advancement advancement = this.build(id, fileHelper);
++         consumer.accept(advancement);
++         return advancement;
++      }
++
++      public Advancement build(ResourceLocation id, net.minecraftforge.common.data.ExistingFileHelper fileHelper) {
++         if(!net.minecraftforge.common.data.AdvancementBuilderHelper.build(fileHelper, this)) {
++            throw new IllegalStateException("Tried to build incomplete advancement!");
++         } else {
++            if (this.field_192066_f == null) {
++               this.field_192066_f = this.field_199751_g.createRequirements(this.field_192065_e.keySet());
++            }
++            return new Advancement(id, this.field_192062_b, this.field_192063_c, this.field_192064_d, this.field_192065_e, this.field_192066_f);
++         }
++      }
```

Comment:
So I'm going to reiterate Lex's point here. Most of this could be moved into a different class. Technically, all you should need to do here is have a method that takes in the advancement builder, data provider consumer, id, and the existing file helper. You could then call `#canBuild` within that method with the function to create a dummy advancement (which is what you are doing in `AdvancementBuilderHelper#build` and then `Builder#build` with no issues. Most you should need is three lines here, one for the method definition, one for the return statement to the outside method call, and one for the end bracket.